apps/src/weblab/brambleHost.js

```diff
@@ -392,49 +392,62 @@ function onInspectorChanged(callback) {
+    const wrappedCallback = err => {
+      ensureMounted();
+      callback(err);
+    };
 
-    // Sync what's on the service locally (should be nothing, but this ensures
-    // our _lastSyncedVersionId matches before we make changes)
-    syncFiles(err => {
-      if (err) {
-        console.warn(`Bramble host: Initial syncFiles failed with error: ${err}`);
-      }
-      // put the source files into the Bramble file system
-      putFilesInBramble(startSources, err => {
-        if (err) {
-          callback();
-        } else {
-          if (forceResetToStartSources) {
-            console.log('startInitialFileSync: forceResetToStartSources - calling syncFiles()');
-            // Start the initial files sync
-            syncFiles(callback);
-          } else {
-            // First-time level load - no project versionId
+    if (err) {
+      wrappedCallback(err);
+      return;
+    }
 
-            // Ignore all of the change events that just occured as we wrote the
-            // startSources into bramble - there is no need to save these during
-            // sync until a user-initiated change occurs
-            resetBrambleChangesAndProjectVersion(currentProjectVersion);
-            webLab_.registerBeforeFirstWriteHook(uploadAllFilesFromBramble);
+    const currentProjectVersion = webLab_.getCurrentFilesVersionId();
+    if (!currentProjectVersion || forceResetToStartSources) {
+      // Get initial sources to put in file system
+      const startSources = webLab_.getStartSources();
 
-            // Start the initial files sync
-            syncFiles(callback);
-          }
+      // Sync what's on the service locally (should be nothing, but this ensures
+      // our _lastSyncedVersionId matches before we make changes)
+      syncFiles(err => {
+        if (err) {
+          console.warn(`Bramble host: Initial syncFiles failed with error: ${err}`);
         }
+        // put the source files into the Bramble file system
+        putFilesInBramble(startSources, err => {
+          if (!err) {
+            if (forceResetToStartSources) {
+              console.log('startInitialFileSync: forceResetToStartSources - calling syncFiles()');
+            } else {
+              // First-time level load - no project versionId
+
+              // Ignore all of the change events that just occured as we wrote the
+              // startSources into bramble - there is no need to save these during
+              // sync until a user-initiated change occurs
+              resetBrambleChangesAndProjectVersion(currentProjectVersion);
+              webLab_.registerBeforeFirstWriteHook(uploadAllFilesFromBramble);
+            }
+          }
+          // Start the initial files sync
+          syncFiles(wrappedCallback);
+        });
       });
-    });
-  } else {
-    ensureProjectRootDirExists(err => {
-      // Start the initial files sync
-      syncFiles(callback);
-    });
-  }
+    } else {
+      syncFiles(wrappedCallback);
```

Comment:
No need to do right now, but I think all of `startInitialFileSync` might implement much nicer using Promises or async/await.  This pair of `syncFiles(wrappedCallback)` calls smells like a 'finally' case on the end of a Promise chain that may change depending on circumstances.

Reply:
yes, I agree that it could be rewritten with those tools and it would be easier to maintain